### DISCUSSION TITLE
reclaiming space and analyzing - Python

### DISCUSSION THREAD

**Discussion by M2ATrail** [duckdb/duckdb#13645] [Q&A]
I'm using python, duckdb 1.0.0
I do some altering to certain columns and create view and close, I'm creating the db using
`conn.execute(f"CREATE OR REPLACE TABLE 'everflowReporting' AS SELECT * FROM read_csv_auto('{folder_path}\*.csv')")`
if the file doesn't exist, new connection, it it around 300mb at close.
if i run it again, and the file exist, it is around 450mb
running those and closing doesn't reduce size.
```
print(conn.execute('FORCE CHECKPOINT;').fetch_df())
print(conn.execute('VACUUM ANALYZE;').fetch_df())
print(conn.execute('show tables;').fetch_df())
```
the first two are in fact returning success.
I can just run it on memory or delete and recreate file everytime, but just curious what is going wrong here.

**Comment by NickCrews** [comment]
What is the problem? what do you want to happen? an extra 150Mb doesn't seem like the end of the world to me, depending on the context. But it might be a problem in others. Running in memory sounds like a great workaround. Is there a reason that isn't ideal?

Can you provide the data (or some dummy data) and a reproducible script?

**Comment by M2ATrail** [comment]
@NickCrews 
I sadly can't provide the data, but I can attach the table schema, sadly some data types are a problem with the source (for example, revenue sometimes in negative so that's why it defaulted to double.)

it isn't the 150MB that's the problem, it is 50% extra space taken for no reason and the commands not working.
I have other things I'm considering, like over a million text files that I want to store meta data needed + the textfile itself, we're talking about over a hundred gigabyte of raw text, don't want that to happen there.

I just tested, exported and imported the db, it is actually double the size.
![image](https://github.com/user-attachments/assets/4384b0d1-851d-4dc0-aa07-6260ce00ab71)

I've also ran into error with this and went to `schema.sql` and commented out the schema create.
also while at it, why is the default export csv not parquet?
```
PS D:\MYPATH> duckdb my_database.db -c "EXPORT DATABASE 'tmp'" && duckdb my_database.new.db -c "IMPORT DATABASE 'tmp'"
100% ?████████████████████████████████████████████████████████████?
Catalog Error: Schema with name "information_schema" already exists!
PS D:\MYPATH> duckdb my_database.db -c "EXPORT DATABASE 'tmp'"
100% ?████████████████████████████████████████████████████████████?
PS D:\MYPATH> duckdb my_database.new.db -c "IMPORT DATABASE 'tmp'"

Catalog Error: Schema with name "information_schema" already exists!
PS D:\MYPATH> duckdb my_database.new.db -c "IMPORT DATABASE 'tmp'"
Catalog Error: Cannot drop entry "information_schema" because it is an internal system entry
PS D:\MYPATH> duckdb my_database.new.db -c "IMPORT DATABASE 'tmp'"
                                                                                                                       Catalog Error: Schema with name "pg_catalog" already exists!
PS D:\MYPATH> duckdb my_database.new.db -c "IMPORT DATABASE 'tmp'"

100% ?████████████████████████████████████████████████████████████?
PS D:\MYPATH>
```


Table:
```
CREATE TABLE everflowReporting(
    date DATE,
    network_offer_id BIGINT,
    offer_name VARCHAR,
    network_advertiser_id BIGINT,
    advertiser_name VARCHAR,
    network_offer_creative_id BIGINT,
    creative_name VARCHAR,
    network_affiliate_id BIGINT,
    affiliate_name VARCHAR,
    sub3 VARCHAR,
    sub5 VARCHAR,
    impressions BIGINT,
    gross_clicks BIGINT,
    total_clicks BIGINT,
    unique_clicks BIGINT,
    duplicate_clicks BIGINT,
    invalid_clicks BIGINT,
    total_conversions BIGINT,
    CV BIGINT,
    invalid_conversions_scrub BIGINT,
    view_through_conversions BIGINT,
    events BIGINT,
    view_through_events BIGINT,
    CVR DOUBLE,
    EVR DOUBLE,
    CTR DOUBLE,
    CPC DOUBLE,
    CPA DOUBLE,
    EPC DOUBLE,
    RPC DOUBLE,
    RPA DOUBLE,
    payout DOUBLE,
    revenue DOUBLE,
    margin DOUBLE,
    profit DOUBLE,
    gross_sales DOUBLE,
    ROAS DOUBLE,
    gross_sales_vt DOUBLE,
    RPM DOUBLE,
    CPM DOUBLE,
    avg_sale_value DOUBLE,
    media_buying_cost DOUBLE
);

**Reply by NickCrews** [reply]
It looks like your DB is only a few hundred kb, not MB? I bet vacuum only works when over a certain size threshold, ie it operates on row groups of 1M rows or something. So can you go up to the scale you actually are worried about (just using dummy data or repeating that input data you have over and over), try the vacuum commands there, and see if it actually is still a problem?

**Reply by NickCrews** [reply]
Also, I'm a community member and not a maintainer. So it might actually be a bug.

**Reply by M2ATrail** [reply]
@NickCrews Wakey wakey bruh, that's in hunderds of thousands.

Yeah I thought it is a bug, but posted here thinking maybe I'm dumb or something and using it wrong.

unrelated but does github notify you by default when someone replies? because it doesn't for me.

**Reply by NickCrews** [reply]
Oops, yup just saw the kb and assumed.

I AM curious if it persists at larger scales, if you get an answer it would be great to hear your followup.

GH does send me a notification even without a tag, idk that might just be a setting. @M2ATrail

**Reply by M2ATrail** [reply]
Well went back to check settings, it was on for email.

sadly the large scale ones don't have much transformations to do, I tried filling an issue but it requires too much info lol.
hopefully one of the maintainers can get to this.

**Comment by Tishj** [comment]
This sounds similar to https://github.com/duckdb/duckdb/issues/13037 which we're still investigating

Ah no I stand corrected, the memory portion looks related, the increased size is an expected limitation currently, due to the inability to reclaim space in the middle of a db file. There are blocks at higher indices preventing the reclaiming of space

**Reply by M2ATrail** [reply]
Alright, so the only option is just export and import?
I've millions of text files (json, HTML) that I want to store some meta data for efficiently before zipping for archiving, does this mean duckdb might not be a good option for me here?

Also why export import uses CSV instead of something more efficient like parquet? I guess csv would be less memory heavy but disk heavy?

**Reply by Tishj** [reply]
That's just the default, parquet is possible for export/import database, see the docs

**Reply by M2ATrail** [reply]
Yeah, I just found it odd that you'd go with CSV as default.
do you happen to know what was causing the error in the export import? first code block https://github.com/duckdb/duckdb/discussions/13645#discussioncomment-10506627

**Comment by Mytherin** [comment]
Thanks for the report!

As mentioned in the [VACUUM](https://duckdb.org/docs/sql/statements/vacuum.html) documentation, `VACUUM` is provided as a compatibility shim for Postgres programs. `VACUUM ANALYZE` recomputes table statistics. `VACUUM` currently does not yet do anything to reclaim space.

The space is not lost - but rather present in free blocks. You can confirm this with the `PRAGMA database_size` command. Essentially due to performing a `DROP` and a `CREATE` in the same transaction (as a result of `CREATE OR REPLACE`) the system cannot overwrite the blocks of the original table with the data of the new table - as that might result in data loss in case of an outage. Instead, the system must first write the data of the new table, and then remove the data of the old table. Since the old table is located in the middle of the file, space cannot be reclaimed. The blocks are instead marked as free and re-used when new data is written.

DuckDB currently has no single command that can be used to compact the file. You can manually run this compaction by attaching a new file and copying over the data using [COPY FROM DATABASE](https://duckdb.org/docs/sql/statements/copy.html#copy-from-database--to), e.g.:

```sql
ATTACH 'new_file.db' AS new_file;
COPY FROM DATABASE current_file TO new_file;
```

Alternatively you can use `EXPORT/IMPORT` as described - or attach a new database instead of overwriting an existing table to begin with.
